File: results/1.0/Bitcoin-transactions/bitcoin-transactions-claude-3-5-haiku-20241022.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <hammer/hammer.h>

typedef struct {
    HBytes* previous_tx_hash;
    uint32_t previous_output_index;
    HBytes* script_sig;
    uint32_t sequence_number;
} TransactionInput;

typedef struct {
    uint64_t amount;
    HBytes* script_pubkey;
} TransactionOutput;

typedef struct {
    uint32_t version;
    HPtrArray* inputs;
    HPtrArray* outputs;
    uint32_t locktime;
    
    uint8_t marker;
    uint8_t flag;
    HPtrArray* witness_data;
} BitcoinTransaction;

static HParser* previous_tx_hash_parser() {
    return h_repeat_n(h_ch(0), 32);
}

static HParser* varint_parser() {
    return h_choice(
        h_uint8(),
        h_uint16(),
        h_uint32(),
        h_uint64(),
        NULL
    );
}

static HParser* script_parser() {
    return h_length_value(varint_parser(), h_many(h_ch(0)));
}

static HParser* transaction_input_parser() {
    return h_sequence(
        previous_tx_hash_parser(),
        h_uint32(),
        script_parser(),
        h_uint32(),
        NULL
    );
}

static HParser* transaction_output_parser() {
    return h_sequence(
        h_uint64(),
        script_parser(),
        NULL
    );
}

static HParser* bitcoin_transaction_parser() {
    HParser* standard_parser = h_sequence(
        h_uint32(),
        h_length_value(varint_parser(), transaction_input_parser()),
        h_length_value(varint_parser(), transaction_output_parser()),
        h_uint32(),
        NULL
    );
    
    HParser* segwit_parser = h_sequence(
        h_ch(0x00),
        h_ch(0x01),
        h_uint32(),
        h_length_value(varint_parser(), transaction_input_parser()),
        h_length_value(varint_parser(), transaction_output_parser()),
        h_length_value(varint_parser(), h_length_value(varint_parser(), h_ch(0))),
        h_uint32(),
        NULL
    );
    
    return h_choice(standard_parser, segwit_parser, NULL);
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <transaction_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation failed");
        fclose(file);
        return 1;
    }

    if (fread(buffer, 1, file_size, file) != file_size) {
        perror("Failed to read file");
        free(buffer);
        fclose(file);
        return 1;
    }
    fclose(file);

    HParser* parser = bitcoin_transaction_parser();
    HParseResult* result = h_parse(parser, buffer, file_size);

    if (result && result->ast) {
        printf("Transaction parsed successfully\n");
    } else {
        printf("Transaction parsing failed\n");
    }

    free(buffer);
    h_parse_result_free(result);
    h_parser_free(parser);

    return 0;
}

File: results/1.0/Bitcoin-transactions/bitcoin-transactions-claude-3-5-haiku-20241022.rs
use std::fs::File;
use std::io::Read;
use std::env;
use nom::{
    multi::{count},
    number::complete::{le_u32, le_u64, le_u8},
    bytes::complete::{take},
    sequence::{tuple},
    combinator::{map, map_opt},
    IResult,
};

#[derive(Debug)]
struct BitcoinTransaction {
    version: u32,
    inputs: Vec<TransactionInput>,
    outputs: Vec<TransactionOutput>,
    locktime: u32,
}

#[derive(Debug)]
struct TransactionInput {
    previous_tx_hash: [u8; 32],
    previous_output_index: u32,
    script_sig_length: u8,
    script_sig: Vec<u8>,
    sequence: u32,
}

#[derive(Debug)]
struct TransactionOutput {
    amount: u64,
    script_pubkey_length: u8,
    script_pubkey: Vec<u8>,
}

fn parse_transaction_input(input: &[u8]) -> IResult<&[u8], TransactionInput> {
    map(
        tuple((
            take(32usize),
            le_u32,
            le_u8,
            map_opt(|i| take_bytes(i), |bytes: Vec<u8>| Some(bytes)),
            le_u32
        )),
        |(hash, prev_index, script_sig_len, script_sig, sequence)| TransactionInput {
            previous_tx_hash: hash.try_into().unwrap(),
            previous_output_index: prev_index,
            script_sig_length: script_sig_len,
            script_sig,
            sequence,
        }
    )(input)
}

fn take_bytes(input: &[u8]) -> IResult<&[u8], Vec<u8>> {
    let (remaining, length) = le_u8(input)?;
    map(take(length as usize), |bytes: &[u8]| bytes.to_vec())(remaining)
}

fn parse_transaction_output(input: &[u8]) -> IResult<&[u8], TransactionOutput> {
    map(
        tuple((
            le_u64,
            le_u8,
            map_opt(|i| take_bytes(i), |bytes: Vec<u8>| Some(bytes))
        )),
        |(amount, script_pubkey_len, script_pubkey)| TransactionOutput {
            amount,
            script_pubkey_length: script_pubkey_len,
            script_pubkey,
        }
    )(input)
}

fn parse_bitcoin_transaction(input: &[u8]) -> IResult<&[u8], BitcoinTransaction> {
    let (input, version) = le_u32(input)?;
    let (input, input_count) = le_u8(input)?;
    let (input, inputs) = count(parse_transaction_input, input_count as usize)(input)?;
    let (input, output_count) = le_u8(input)?;
    let (input, outputs) = count(parse_transaction_output, output_count as usize)(input)?;
    let (input, locktime) = le_u32(input)?;

    Ok((input, BitcoinTransaction {
        version,
        inputs,
        outputs,
        locktime,
    }))
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <transaction_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_bitcoin_transaction(&buffer) {
        Ok((_, transaction)) => {
            println!("Parsed Transaction: {:?}", transaction);
            Ok(())
        },
        Err(e) => {
            eprintln!("Parsing error: {:?}", e);
            std::process::exit(1)
        }
    }
}